backend/wmg/data/load_corpus.py

```diff
@@ -177,43 +182,93 @@ def get_X_raw(anndata_object: anndata.AnnData) -> Union[np.ndarray, sparse.spmat
-                coo_sparse_raw_expression_matrix.col, rankit_normalized_coo_sparse_raw_expression_matrix.col
+            raw_expression_csr_matrix = sparse.csr_matrix(expression_matrix[start:end, :])
+
+            # Compute RankIt
+            rankit_integrated_csr_matrix = rankit(raw_expression_csr_matrix)
+
+            rankit_integrated_coo_matrix = filter_out_rankits_with_low_expression_counts(
+                rankit_integrated_csr_matrix, raw_expression_csr_matrix, expect_majority_filtered=True
             )
-            rankit_data = rankit_normalized_coo_sparse_raw_expression_matrix.data
 
-            array[rows, cols] = {"data": raw_data, "rankit": rankit_data}
+            global_rows = rankit_integrated_coo_matrix.row + start + first_obs_idx
+            global_cols = global_var_index[rankit_integrated_coo_matrix.col]
+
+            rankit_data = rankit_integrated_coo_matrix.data
+
+            assert len(rankit_data) == len(global_rows)
+            assert len(rankit_data) == len(global_cols)
+
+            array[global_rows, global_cols] = {"rankit": rankit_data}
             del (
-                coo_sparse_raw_expression_matrix,
-                rankit_normalized_coo_sparse_raw_expression_matrix,
-                rows,
-                cols,
-                raw_data,
+                raw_expression_csr_matrix,
+                rankit_integrated_coo_matrix,
+                rankit_integrated_csr_matrix,
+                global_rows,
+                global_cols,
                 rankit_data,
             )
             gc.collect()
 
     logger.debug(f"Saved {array_name}.")
+
+
+def filter_out_rankits_with_low_expression_counts(
+    rankits: csr_matrix, raw_counts_csr: csr_matrix, expect_majority_filtered=True
+) -> coo_matrix:
+    """
+    Keep only rankit values that were computed from expression values above the desired threshold.
+
+    @param expect_majority_filtered: Set this to True if the caller expects the majority of rankit values will be
+    filtered, as we can then use an optimal implementation
+    """
+
+    rankits_nnz_orig = rankits.nnz
+    raw_counts = raw_counts_csr.tocoo(copy=False)
+    to_keep_mask = raw_counts.data > RANKIT_RAW_EXPR_COUNT_FILTERING_MIN_THRESHOLD
+
+    # Unfortunately, it seems to take longer to decide which strategy to use than it does to just run either one.
+    # start = time.time()
+    # expect_majority_filtered = to_keep_mask.sum() < rankits_nnz_orig / 2
+    # end = time.time()
+    # logger.info(f"filter_out_rankits_with_low_expression_counts(): use extract strategy={expect_majority_filtered}; "
+    #             f"decision time={end - start}")
+
+    start = time.time()
+
+    if expect_majority_filtered:
+        rows_to_keep = raw_counts.row[to_keep_mask]
+        cols_to_keep = raw_counts.col[to_keep_mask]
+        rankits_to_keep = rankits.data[to_keep_mask]
+        rankits_filtered = coo_matrix((rankits_to_keep, (rows_to_keep, cols_to_keep)))
```

Comment:
@ebezzi This logic is what's now being used to perform the rankit filtering. I ran an ad hoc benchmark on a 1.3GB H5AD with ~91M expr values and it took ~1 sec to perform the filtering down to ~18M values. Unless there's some other data "profile" that I'm missing (dense matrix h5ads?), the performance is good. Let me know if you agree since you spent some time on understanding the optimizations needed for the earlier implementation. Thanks!

Reply:
This seems pretty similar to the code in the else branch, so this looks good to me. The "naive" solution `sparse_matrix[sparse_matrix < 3]` would take several minutes and raise a warning (something like "comparison on sparse matrices should be avoided), so I believe this is fine.

Reply:
The `eliminate_zeros()` was also needed for restoring the sparsity of the matrix, which I assume is equivalent to line 256 here.

Reply:
yup, exactly! thank you for reviewing!